Feature/lego 21/modifications following review

abulford opened the pull request.

abulford committed:
Removing migrator user and home directory

This setup with the migrator user and with its home directory being a volume which was also the workspace meant that an SSH agent couldn't be mounted in to the container to allow SSH use without providing a private key.

abulford committed:
Removing entrypoint and wait-up scripts

This is changing the image from being something which is run as a long-lived container which can be exec'd into, to one which can be run with the working directory mounted and removed once the current job is finished, as discussed.  By keeping state in a mounted directory on the host, a  lot of the complication in the wrapper scripts isn't needed so can be removed.

abulford committed:
Adding get-authors.sh script

This can be used to populate an 'authors.txt' file with username mappings from SVN to Git world.

abulford committed:
Adding docker-compose.yml

This allows easy running of the container with the host's SSH agent mounted in, removing the need to mount your private key into the container.

abulford committed:
Minor tweak to fix comments

abulford committed:
Changing argument order in get-authors.sh

Removing redmatter.com as the default email domain to make this script more generic.

abulford committed:
Updating README.md

codemedic commented on get-authors.sh:
```diff
@@ -0,0 +1,43 @@
+REPO_URL=$1
+EMAIL_DOMAIN=$2
+OUTPUT_PATH=${3:-authors.txt}
+
```
Looks good, but you could have used an XML processor to make it  tidy.

`xmlstarlet` is a sorta swiss army knife for XML processing

```
apt-get install xmlstarlet
```

Then the below command would habe done the job.

```
svn log -l10 --xml | \
xmlstarlet sel -t -v '//author/text()' --nl - | \
sort | uniq | while read u; do \
    echo "$u = $(echo $u | tr . ' ' | sed 's/\([a-z]\)\([a-zA-Z0-9]*\)/\u\1\2/g') <$u@redmatter.com>"; \
done
```

The formula above is roughly based on your logic; main difference though is to pre-process the name from `svn log` to remove the dot  if there is any, before capitalising the first letter of name components.

abulford replied:
Nice!  I've updated the script to use your much neater version, with a few tweaks to make it a little easier (for me) to read.  I've also added usage information.

codemedic commented on entrypoint.sh:
```diff
@@ -1,59 +0,0 @@
-        fi
-        ;;
-
-    wait-up)
```
Migrating a small project will be quick and in most cases this docker image can be run as a "one-shot" command. In cases where the project is huge, and if it has to run over a long period of time, say over a weekend or something, then it will be handy to be able to exec into a running container and setup the migration, with the help of `screen` and then detach from it.

Thats the way it was originally envisaged. Now that was with the limited understanding of docker and its way of working back then. Very likely that it is not the right approach in docker land.

Obviously those aspects are the concern of the host machine; but what if it doesnt come with the right tools for it; i.e if it is CoreOS, which doesn't come with `screen` utility.

Only way I can see this new form working is if it is possible to `docker run` this image from within another container that has the right tools and has access to the host's docker daemon.

As we saw on Friday SZ had trouble with the long running migration. Would have helped if it was possible to run from within LDC.

abulford replied:
I think the Docker-y answer to the question of how to run long running migrations which won't fail if the user disconnects is to detach from the container.  Once you've kicked off the `svn2git` command you can detach with `Ctrl-P + Ctrl-Q`.

Alternatively, you could kick off the migration in detached mode without ever entering the container.  For example, the following few commands get the list of authors, kick off the `svn2git` migration in detached mode then use docker logs to watch the output:

```
docker run --rm --network=host -v ~/.ssh/known_hosts:/root/.ssh/known_hosts -v ${SSH_AUTH_SOCK}:${SSH_AUTH_SOCK} -v ${PWD}/workspace:/workspace -e SSH_AUTH_SOCK=${SSH_AUTH_SOCK} redmatter/svn2git-workspace sh -c "get-authors svn+ssh://user@repo/trunk email.com > authors.txt"
docker run -d --name=my-long-running-git-migration --network=host -v ~/.ssh/known_hosts:/root/.ssh/known_hosts -v ${SSH_AUTH_SOCK}:${SSH_AUTH_SOCK} -v ${PWD}/workspace:/workspace -e SSH_AUTH_SOCK=${SSH_AUTH_SOCK} redmatter/svn2git-workspace sh -c "svn2git svn+ssh://user@repo/trunk --svn2git-args"
docker logs -f my-long-running-git-migration
```

While a lot of that would be needed anyway (the volumes and environment variables), it is a bit long-winded, so potentially could be pulled into the `docker-compose.yml` if we wanted to support this mechanism.  I think simply detaching from the container if you need to is probably the best approach, though.

I don't want to completely change the image if you're not happy with it, so if you really think we need to support `screen` etc I can try to re-instate these files, but I removed them in the first place because the home directory set up in `entrypoint.sh` was interfering with my ability to mount the SSH agent in, so at the time if felt like it was more trouble than it was worth to try and maintain.

codemedic replied:
The change is good; it has made thing tidy and simple.
As long as the use-case is addressed, I am ok with it.

abulford committed:
Improvements to get-authors.sh

- Using xmlstarlet to extract <author> element values
- Changing usage of script to just output the mappings, this may then be redirected to authors.txt
- Making better use of sed and tr to greatly simplify the script and allow any number of dots in the SVN username
- Updated README to reflect changes to script

Improvements were as suggested in @codemedic's comment on #1

abulford committed:
Renaming get-authors.sh to get-authors

abulford committed:
get-authors can take svn log options

This allows the user to pass -l100, for example, to restrict the number of commits checked for authors to the last 100.

abulford committed:
Adding usage information to get-authors
